### PR TITLE
pre-auth and capture

### PR DESCRIPTION
**Pull request by johnnye**
This adds missing methods for pre-auth in GoCardless. As this is an API that hands off to the site it is at the moment 3 new methods. `authorize`, `completeAuthorize` and `capture`. 

This is quite a big PR relative to the size of the gateway so I wanted to get this out asap for review. Happy to work on cleaning up whats needed (see TODOs below). This is also my first foray into Omnipay internals so happy to adapt to style where it is needed. 

Developer documentation for this is available at https://developer.gocardless.com/#pre-authorization although at the moment there is a missing parameter in the docs `pre_authorization [merchant_id]`. This has been raised with GoCardless. 

This also add the option to use the `state` parameter in current requests. `state` can be used to pass around a reference for payment requests as they are handed off site and back again.

### CODE REVIEW

**Comment by johnnye** [comment]
I'll not be including the ability to cancel pre-auths in this PR. The omnipay method `void` seems to logically correspond with cancelling individual payments (bills in a GoCardless world), rather then pre-auths.

File: src/Message/CaptureResponse.php [hunk 1]
```diff
@@ -0,0 +1,30 @@
...
+namespace Omnipay\GoCardless\Message;
+
+use Omnipay\Common\Message\AbstractResponse;
+use Omnipay\Common\Message\RequestInterface;
+
+class CaptureResponse extends AbstractResponse
+{
+    protected $transactionReference;
+
+    public function __construct(RequestInterface $request, $data)
```
Comment (greydnls) [inline src/Message/CaptureResponse.php]: This constructor isn't necessary. It will by default call the parent constructor.

**Comment by greydnls** [comment]
This looks good. If you want to check off your final todo, and then squash some of those commits down I'd be happy to merge it in and tag a new version.

**Comment by johnnye** [comment]
thanks - will do this on the clock tomorrow.

**Comment by johnnye** [comment]
If you want me to squash it all into one commit, please let me know. I also need to update this PR to match issue #6,

**Comment by johnnye** [comment]
Is there anything that I need to get this and #7 accepted? 

Do I need to squash these commits into one commit?

**Comment by benrcole** [comment]
@kayladnls I'm after support for omnipay/gocardless recurring payments, could this PR be merged? ( @johnnye  thanks for the work)

**Comment by greydnls** [comment]
Thanks @johnnye sorry for the delay. I've got this and #3 merged and tagged a new release.

**Pull request merged**